### PR TITLE
BTF support

### PR DESCRIPTION
**Pull request by olsajiri**
hi,
 **PLEASE** take this as RFC.. ;-)

I just wanted to check on BTF support and it turned out
not to be that much code so far to load structs definitions
and get it working with bpftrace code generation.

It's just a middle step, my goal is to have support
to dereference and display probe arguments.

I'll be grateful for any insight on how you guys want
to drive/design this, or if I've already missed some
posts/discussions.

Here's what you need to do to run this:
  - get latest dwarves with pahole v1.13
  
  - install libbpf

  - use latest bcc with renamed libbpf.so to libbcc_bpf.so

  - build kernel with BTF support, got mine from:
    https://kernel.googlesource.com/pub/scm/linux/kernel/git/bpf/bpf-next master

  - enable kernel config options:
    CONFIG_DEBUG_INFO_BTF=y

    to get BTF data generated during compile:
    $ make
    ...
      LD      vmlinux
      BTF     vmlinux
    ...

  - get the BTF data into separate file:
    $ objcopy --dump-section .BTF=./btf ./vmlinux

  - boot the new kernel ;-)

  - run bpftrace with:

    BPFTRACE_BTF=./btf bpftrace -e 'kprobe:do_dentry_open { printf("pid %d ppid %u comm %s\n", \
        ((struct task_struct *) curtask)->pid,
        ((struct task_struct *) curtask)->real_parent->pid,
        ((struct task_struct *) curtask)->comm); }'
      Attaching 1 probe...
      pid 868 ppid 1 comm systemd-journal
      pid 868 ppid 1 comm systemd-journal
      pid 1125 ppid 1 comm NetworkManager
      pid 1125 ppid 1 comm NetworkManager
      pid 1123 ppid 1 comm sssd
      pid 1125 ppid 1 comm NetworkManager
      pid 1125 ppid 1 comm NetworkManager
      pid 1125 ppid 1 comm NetworkManager
    ...

The current idea is to have BTF data placed in kernel rpm,
because it's quite small (check below, kudos to FB guys!),
so above functionality would be available without kernel
debuginfo packages.

  $ objcopy --dump-section .BTF=./btf ./vmlinux
  $ ll vmlinux
  -rwxrwxr-x. 1 jolsa jolsa 6241904288 Apr 24 22:10 vmlinux
  $ ll btf
  -rwxrwxr-x. 1 jolsa jolsa 14805539 Apr 24 22:09 btf

thoughts? thanks,
jirka

### CODE REVIEW

**Comment by brendangregg** [comment]
Wow, thanks for trying this out, that's very interesting! Yes, we want BTF support -- this is great work to get it working without much changes.

@yonghong-song how is BTF coming along?

**Comment by brendangregg** [comment]
And Jiri -- glad you're taking a look at bpftrace :-) things have been coming along rapidly in the last few months.

**Comment by olsajiri** [comment]
yea, I like it, it's real useful ;-)

I just realized you need new bcc and libbpf installed to compile/run it,
I guess that's why the CI tests failed (I updated the main comment)

File: src/ast/ast.h [hunk 1]
```diff
@@ -28,6 +28,7 @@ class Expression : public Node {
   bool is_literal = false;
   bool is_variable = false;
   bool is_map = false;
+  static std::vector<std::string> resolve;
```
Comment (ajor) [inline src/ast/ast.h]: This could be a static member of the Cast class, to keep it more localised

Comment (ajor) [reply src/ast/ast.h]: I like the idea though - I think there's some other code in here we could clean up with more use of static members like this

Comment (olsajiri) [reply src/ast/ast.h]: > This could be a static member of the Cast class, to keep it more localised

right, will move it there

**Comment by olsajiri** [comment]
> Nice!
> 
> I've not looked at it all that closely, but is it feasible to wrap the kernel calls so we can have self contained tests with fixed inputs, like in
> 
> [bpftrace/tests/tracepoint_format_parser.cpp](https://github.com/iovisor/bpftrace/blob/480bb22c3b85dea8152f9668efa12e49b6177b04/tests/tracepoint_format_parser.cpp#L17)
> 
> Line 17 in [480bb22](/iovisor/bpftrace/commit/480bb22c3b85dea8152f9668efa12e49b6177b04)
>  TEST(tracepoint_format_parser, tracepoint_struct)

I'll check on how to add tests for BTF class,

Fedora 30 should have soon bcc 0.9.0 together with the rename of libbpf,
so should be easier to build this and get through CI

**Comment by olsajiri** [comment]
I pushed out new version with following changes:
  - adding libbpf detection
  - adding tests
  - adding detault 'task_struct' cast to curtask builtin
  - Fedora 30 now ships libbpf
  - adding lazy BTF type resolving

**Comment by olsajiri** [comment]
I pushed out new version with following changes:
  - adding libbpf detection
  - adding tests
  - adding detault 'task_struct' cast to curtask builtin
  - Fedora 30 now ships libbpf
  - adding lazy BTF type resolving

**Comment by olsajiri** [comment]
I'll create non-RFC pull request to avoid confusion

**Pull request closed without merging**